packages/block-editor/src/components/block-list/moving-animation.js

```diff
@@ -29,10 +29,17 @@ import { useReducedMotion } from '@wordpress/compose';
  */
 function useMovingAnimation( ref, isSelected, enableAnimation, triggerAnimationOnChange ) {
 	const prefersReducedMotion = useReducedMotion() || ! enableAnimation;
-	const [ resetAnimation, setResetAnimation ] = useState( false );
+	const [ triggeredAnimation, triggerAnimation ] = useReducer( ( state = 0 ) => state + 1 );
```

Comment:
Is useReducer needed here? The logic using a reducer to do a sum is complex to follow, can we use two-state flags (triggeredAnimation, finishedAnimation) and change the flags accordingly?
If using useReducer can we use a single reducer with multiple actions if that makes things more simple?

Reply:
I find it simpler this way personally. Each hook keeps track of the "id" of the animation. Basically, it's just a counter that increments each time you call the dispatch function.

Reply:
Could this be a personal preference. Or maybe it's just because we're not used to using `useReducer` that much?

Reply:
I don't want to discard that feedback thought, what do you think? maybe others could share opinions here. @aduth @epiqueras

Reply:
`useReducer` is useful when you need handlers to read props and current state, but don't want to create a new function on every render or whenever they change if you are memoizing. In this case you could just use `setState`:

```js
const [ state, setState ] = useState( 0 );
setState( ( state ) => ++state );
```

I don't think that makes a big difference though. I do agree with @jorgefilipecosta that it would be nicer if we could just use 2 booleans instead of ever increasing numbers. Something like:

```js
const [ { started, finished }, setAnimationState ] = useState( {
	started: false,
	finished: false,
} );
```

Also note that you should pass `initialState` as the second parameter to `useReducer` instead of using a default parameter in the reducer function, so that it is defined before the first update.

https://github.com/facebook/react/blob/42b75ab007a5e7c159933cfdbf2b6845d89fc7f2/packages/react-reconciler/src/ReactFiberHooks.js#L635

Reply:
Ohhh, I wasn't clear enough. We can't use booleans (it was the previous approach) to fix the race conditions completely. We need to be sure all the animations (multiple rerenders can trigger two in a row) have been triggered before resetting.

> const [ state, setState ] = useState( 0 );
> setState( ( state ) => ++state );

This is actually a bad pattern, it creates unexpected bugs especially if you trigger setState from effects because `state` can be outdated at that point. React team talked about that a little bit. The idea is that each time the updated state value depends on the previous state value, it's better to use `useReducer`.

Reply:
I extracted the reducer to avoid recreating a function on each render, I also used the default value argument.

Reply:
>Ohhh, I wasn't clear enough. We can't use booleans (it was the previous approach) to fix the race conditions completely. We need to be sure all the animations (multiple rerenders can trigger two in a row) have been triggered before resetting.

You could achieve that with 2 booleans instead of 1 like the previous approach. But, I can see why numbers make this easier.

>This is actually a bad pattern, it creates unexpected bugs especially if you trigger setState from effects because state can be outdated at that point. React team talked about that a little bit. The idea is that each time the updated state value depends on the previous state value, it's better to use useReducer.

That's why my example uses the functional variant of `setState`. I am not closing over the state, React passes the latest value as a param and I operate on that.

>I extracted the reducer to avoid recreating a function on each render, I also used the default value argument.

:rocket: